### PR TITLE
New `Update-ProjectProperties` cmdlet

### PR DESCRIPTION
**Pull request by ChrisLambrou**
Adds a new `Update-ProjectProperties` cmdlet, that can be used to insert or update the `Version`, `AssemblyVersion`, `FileVersion` and `PackageReleaseNotes` properties within a VS2017-style project file. This is to complement `Update-AssemblyVersion` cmdlet, which does a similar job for `AssemblyInfo.cs` files.

For example, given the following project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
   ...etc...
</Project>
```

If you invoke the following:

```powershell
Update-ProjectProperties `
    -Version '1.2.3.4-SomeBranch' `
    -AssemblyVersion '1.0.0.0' `
    -FileVersion '1.2.3.4' `
    -PackageReleaseNotes 'RELEASE-NOTES'
```

The project file's contents will be modified as follows:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <Version>1.2.3.4-SomeBranch</Version>
    <AssemblyVersion>1.0.0.0</AssemblyVersion>
    <FileVersion>1.2.3.4</FileVersion>
    <PackageReleaseNotes>RELEASE-NOTES</PackageReleaseNotes>
  </PropertyGroup>
   ...etc...
</Project>
```

### CODE REVIEW

**Comment by PeterGerrard** [comment]
Given it is now possible to set these values when calling msbuild do we actually want to modify the files at all? Not doing so removes local modifications

You can do `msbuild.exe foo.sln /p:VersionPrefix=1.2.3.4 /p:VersionSuffix=beta1` to get this for versions. (And similar for dotnet tooling)

Not 100% sure if there is one for PackageReleaseNotes but I assume so.

**Comment by ChrisLambrou** [comment]
I looked at specifying properties via the command-line, and it's probably a better approach for most use cases. I ran into problems because those properties are effectively global, applying to all Projects in a Solution. I've hit a scenario where I want to specify different version numbers and package release notes for different projects, and I've not been able to find a way to do that via the msbuild cli.

**Comment by ChrisLambrou** [comment]
At the very least, I should probably highlight in the cmdlet description that it's probably better to use msbuild properties rather than this cmdlet, and explain the more narrow use case for this cmdlet.

File: Public/Update-ProjectProperties.ps1 [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        [Parameter(Mandatory = $False)]
+        [string] $PackageReleaseNotes
+    )
+
+    if (-not (Test-Path $Path)) {
+        throw "Project file not found: $Path"
+    }
+
+    $ProjectXml = [xml] (Get-Content $Path)
+    $PropertyGroupElement = $ProjectXml.SelectSingleNode('/Project[@Sdk]/PropertyGroup[1]')
```
[the remark this answers is not in the document]
Comment (ChrisLambrou) [reply Public/Update-ProjectProperties.ps1]: I tried it, and it ends up being even more verbose than the direct calls against the native .NET objects. Not sure whether or not to keep the change...

File: Public/Update-ProjectProperties.ps1 [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+{
+    [CmdletBinding()]
+    param (
+        [Xml.XmlDocument] $XmlDocument,
+        [Xml.XmlElement] $ParentElement,
+        [string] $PropertyName,
+        [string] $PropertyValue
+    )
+
+    if (-not [string]::IsNullOrEmpty($PropertyValue)) {
```
[the remark this answers is not in the document]
Comment (ChrisLambrou) [reply Public/Update-ProjectProperties.ps1]: Oh, this went through several iterations as I tried to see how I might get this to also support specifying empty values before I concluded it was fruitless. I finally settled on this as something that finally worked, and didn't even think to see if I could simplify it afterwards. D'oh!

**Pull request merged**